include/revng/Support/ManagedStaticRegistry.h

```diff
@@ -0,0 +1,57 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <compare>
+#include <set>
+
+#include "llvm/ADT/Twine.h"
+#include "llvm/Support/ManagedStatic.h"
+
+#include "revng/Support/Debug.h"
+
+template<typename Type>
+concept IsStaticallyRegisterable = requires(Type &&Value) {
+  { Value.key() } -> std::convertible_to<llvm::StringRef>;
+};
+
+template<IsStaticallyRegisterable Registered>
+class RegisterManagedStaticImpl {
```

Comment:
1. Can we call this just `Registry` or `GenericRegistry` or `StaticRegistry`?
2. `s/Registered/RegistryEntry/`
3. I think there's little point in giving the user freedom on the key. Let's force it to `std::string` or `llvm::StringRef`. It's better to have consistent practices than being super flexible.
   We can also get rid of the transparent key comparator.

Reply:
3. I misread the code.
   I think my comment would be: do we really want a `std::set` + `TransparentKeyComparator` over a `std::map`?
   But no biggie.

Reply:
3. IMO I'd still want a transparent comparator even with a map - strings in C++ suck. But no strong preferences on set vs map (they are practically the same thing except a `pair` is hard-coded in one of them)

Reply:
OK. I see the transparent hashing thing with strings is a recurring thing in your code.

Do you think we can abstract this in some reusable "StringMap" data structure (actually, we should also look into `llvm::StringMap` and see if we should start adopting that instead).

Reply:
The problem with none-standard data structures is that debugging them sucks. And simple 
a transparent comparator helps (bonus points for also using flat map we're getting soon-ish)

That being said I considered adding something like
```cpp
namespace detail { struct TransparentComparator; struct TransparentHasher; }

template <typename T>
using OurStringMap = std::map<std::string, T, detail::TransparentComparator>;
template <typename T>
using OurUnorderedStringMap = std::unordered_map<std::string, T, detail::TransparentHasher>;
```

but honestly, I couldn't find enough users to justify it. Especially considering that often I want something "more" out of the comparator, say for it to use `key()` like in this case.